book/forms.md

```diff
@@ -164,102 +164,147 @@ def input(self, node):
-with the address bar, clicking on an input should clear its
-content:[^unless-cursor]
+But keyboard input is harder. Think back to how we [implemented the
+address bar](chrome.md): we added a `focus` field that remembered what
+we clicked on and sent it key presses. We'll want something like that
+for text entries---a `focus` field---but it's going to be more complex
+because text entries live inside a `Tab`, not inside the `Browser`.
 
-[^unless-cursor]: Unless you've implemented some basic editing
-    controls, like the "Cursor" exercise in [Chapter 7](chrome.md).
+Naturally, we will need a `focus` field on each `Tab`, to remember
+which text entry (if any) we've recently clicked on:
 
-``` {.python indent=12}
-elif elt.tag == "input":
-    elt.attributes["value"] = ""
+``` {.python}
+class Tab:
+    def __init__(self):
+        # ...
+        self.focus = None
 ```
 
-Next, typing on the keyboard needs to change the value, and in order
-to do that, we need to set the `focus` to point to this element:
+Now when we click on an input element, we need to set `focus`:
 
-``` {.python indent=12}
-elif elt.tag == "input":
-    # ...
-    self.focus = obj
+``` {.python}
+class Tab:
+    def click(self, x, y):
+        while elt:
+            elif elt.tag == "input":
+                # ...
+                self.focus = elt
+                return
 ```
 
-Until now, the `focus` field was either `None` (nothing has been
-clicked on) or `"address bar"` (the address bar has been clicked on).
-Now we're adding the additional possibility that it is a layout object
-that the user is typing into.
+But remember that keyboard input isn't handled by the `Tab`---it is
+handled by the `Browser`. So how does the `Browser` even know that the
+`Tab` should handle a certain keyboard event? Well, the answer is that
+the `Browser` has to remember that the `Tab` is in focus!
 
-Finally, since we've changed the content of the input element (by
-clearing it) we need to redraw the screen. But unlike before, where a
-simple `draw` call was enough, we're now changing the web page HTML
-itself! This means we must change the layout tree, and to do that, we
-must call `layout`:
+So, when you click on the web page, the `Browser` updates its focus as
+well:
 
-``` {.python indent=12}
-elif elt.tag == "input":
-    # ...
-    self.layout(self.document.node)
+``` {.python}
+class Browser:
+    def handle_click(self, e):
+        if e.y < CHROME_PX:
+            self.focus = None
+            # ...
+        else:
+            self.focus = "content"
+            # ...
+        self.draw()
 ```
 
-Once focus has been moved to an input element, typing on the keyboard
-has to change the input's contents:
+Now when a key press happens, the `Browser` can either send it to the
+address bar or the active tab:
 
-``` {.python indent=4}
-def keypress(self, e):
-    # ...
+``` {.python}
+class Browser:
+    def handle_key(self, e):
+        # ...
+        elif self.focus == "content":
```

Comment:
When changing tabs, you'll have to set self.focus to None

Reply:
I'm not sure I understand this comment, could you elaborate?

Switching tabs happens in `handle_click`, and that sets `Browser.focus` to `None` (see paragraph starting "So, when you click on the web page"). So when you switch tabs keyboard input now goes nowhere.

Or do you mean—when we switch tabs we need to blur the tab we left? As in, do you mean we need to set `Tab.focus` to `None` when we switch tabs? This is a good point—you can switch tabs and the `Tab` has focus somewhere but the `Browser` doesn't have the content focused so typing doesn't go anywhere. I think this is kind of analogous to clicking in the address bar—you can end up with something thinking it's focused when it's not, which is confusing. But at least keyboard entry will go to the right one, and it avoids explicit `blur` calls, which can be moved to an exercise. Does that trade sound bad to you?

Reply:
I see. Code looks ok as-is then.